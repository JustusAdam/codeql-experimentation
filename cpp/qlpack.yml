name: local/query-tests
dependencies:
  local/query: ${workspace}
extractor: cpp