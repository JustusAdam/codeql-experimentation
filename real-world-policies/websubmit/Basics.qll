private import cpp
private import semmle.code.cpp.dataflow.new.TaintTracking

predicate is_sensitive(Type t) {
  exists(string s |
    t.getName() = s and
    (
      s = "ApiKey" or
      s = "LectureQuestionSubmission" or
      s = "LectureQuestion" or
      s = "LectureAnswer"
    )
  )
}

predicate contains_sensitive_type(Type t) {
  exists(Type s |
    is_sensitive(s) and
    t.resolveTypedefs().refersTo(s)
  )
}

predicate is_store_function(Function f) {
  (
    f.getName() = "replace" or
    f.getName() = "insert"
  ) and
  f.getDeclaringType().getName() = "MySqlBackend"
}

predicate is_store(DataFlow::Node n) {
  exists(Variable v |
    n.asVariable() = v and
    is_store_function(v.getEnclosingElement().(Function)) and
    v.getName() = "reval"
  )
  //   exists(Parameter p |
  //     n.asParameter() = p and p.getIndex() = 1 and is_store_function(p.getFunction())
  //   )
}

// predicate is_store(Expr e) {
//   exists(FunctionCall c |
//     (
//       c.getTarget().getName() = "replace" or
//       c.getTarget().getName() = "insert"
//     ) and
//     c.getArgument(1) = e
//   )
// }
module StoreTaintConfig implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node n) {
    any()
    //exists(Type t | is_sensitive(t) and n.getType().refersTo(t))
  }

  predicate isSink(DataFlow::Node n) {
    any()
    //is_store(n.asExpr())
  }
}

module StoreTaint = TaintTracking::Global<StoreTaintConfig>;

module AllFlowsConfig implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node n) { any() }

  predicate isSink(DataFlow::Node n) { any() }
}

module AllFlows = TaintTracking::Global<AllFlowsConfig>;
