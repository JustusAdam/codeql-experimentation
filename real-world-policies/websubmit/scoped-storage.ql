import Basics
import cpp
import semmle.code.cpp.dataflow.new.TaintTracking

from DataFlow::Node n, DataFlow::Node store, Type sens
where
  is_sensitive(sens) and
  n.getType().refersTo(sens) and
  is_store(store) and
  StoreTaint::flow(n, store)
select n, store, n.getType(), store.getLocation()
//
// from Call c, DataFlow::Node node
// where node.asExpr() = c.getArgument(1) and is_store_function(c.getTarget())
// select c, node, node.getLocation()
//
// from Call c, Expr argument, DataFlow::Node n
// where
//   is_store_function(c.getTarget()) and
//   argument = c.getArgument(1) and
//   argument = n.asExpr()
// select c, argument, n, c.getLocation()
