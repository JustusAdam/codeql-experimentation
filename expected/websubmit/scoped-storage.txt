| n | store | col2 | col3 |
+---+-------+------+------+
